Add additional pool exclusive-non-isolcpus

phillyb7 opened the pull request:
Add an additional pool called exclusive-non-isolcpus. This pool is used with cores that are to be isolated from other pods in the cluster but are not allocated as isolcpus. The PR adds a new flag _--excl-non-isolcpus=\<list\>_ which has a default value of "-1".

If the value passed is "-1" then CMK does not get configured with this extra pool and functions as normal.

If a list of cores or a range of cores are given then CMK is configured with these cores placed in the exclusive-non-isolcpus pool

przemeklal commented on cmk.py:
```diff
@@ -79,6 +81,8 @@
   --namespace=<name>           Set the namespace to deploy pods to during the
                                cluster-init deployment process.
                                [default: default].
+  --excl-non-isolcpus=<list>   List of cores to be added to the extra exclusive
```
nit: it's a list of *physical* cores - passing logical cores that exist will throw an error anyway

przemeklal commented on intel/init.py:
```diff
@@ -284,6 +290,30 @@ def check_sst_cp_isolated_cores(platform, num_exclusive_cores,
+            for cpu in cpu_list:
+                if not cpu.isdigit():
+                    raise RuntimeError("Invalid core ID: {}".format(cpu))
+        cpus = topology.parse_cpus_from_isolcpus(cpus_str)
```
nit: this function probably should change its name now :)

przemeklal commented on intel/init.py:
```diff
@@ -309,6 +339,51 @@ def assign(cores, pool, count=None):
+        if c >= len(cores):
+            cores_not_on_system.append(c)
+    if len(cores_not_on_system) > 0:
+        raise RuntimeError("Following cores not on system: {}"
```
nit: as above - if someone (like myself :wink:) passes logical cores it'll throw an error that cores aren't on the system, while they're listed everywhere, I'd leave like is and inform the user that they need to pass physical core, but otherwise it's alright

phillyb7 commented:
Added a clearer explanation about the physical cores being supplied in cmk.py and the error message for cores that are not on the system.

I also added an example for the use of the new pool to the resources folder and changed the name of the _parse_cpus_from_isolcpus_ function to simply _parse_cpus_str_. 

If you think that name is appropriate and there's nothing else to be added I can squash the commits and get it merged :)

przemeklal commented:
@phillyb7 Good stuff! +1 from me and let's merge it
